### PR TITLE
SAMZA-2248: Fix AM bookkeeping on receiving dead container notifications

### PR DESCRIPTION
**Pull request by xinyuiscool**
Fixed the following issue:

1. AM gets extra containers from the RM which it saves for later use
2. When the container that we saved in step1 dies, the AM on receiving this callback does nothing about it.
3. Later, when we are looking for a container to use - we pick up the dead container that we saved and did not clean up (step 1&2) and launch a container. 
4. Now, if this launched container ever dies - the RM will never notify the AM about it since it see's it as a duplicate (step 2)
5. Job is left without the container rescheduled and will need to be restarted.

### CODE REVIEW

**Comment by xinyuiscool** [comment]
@abhishekshivanna  and @vjagadish1989 : please help take a look at this pr. Thx.

**Comment by abhishekshivanna** [comment]
LGTM. Thanks for the fix @xinyuiscool

**Comment by vjagadish1989** [comment]
> (2) When the container that we saved in step1 dies, the AM on receiving this callback does nothing about it.

@xinyuiscool: 
what's the root-cause for (2)? the expected behavior to request resources for it from the ClusterManager and restart it

File: samza-core/src/test/java/org/apache/samza/clustermanager/TestContainerRequestState.java [hunk 1]
```diff
@@ -193,4 +193,28 @@ public void testContainerAssignment() throws Exception {
 
   }
 
+  @Test
```
Comment (rmatharu) [inline samza-core/src/test/java/org/apache/samza/clustermanager/TestContainerRequestState.java]: Could we also add a test for this in TestContainerAllocator and/or TestHostAwareContainerAllocator?

Comment (xinyuiscool) [reply samza-core/src/test/java/org/apache/samza/clustermanager/TestContainerRequestState.java]: I think the scope of the change is quite limited so this test should cover it. The change does not affect the allocator.

File: samza-core/src/main/java/org/apache/samza/clustermanager/ResourceRequestState.java [hunk 2]
```diff
@@ -229,6 +230,25 @@ public int releaseExtraResources() {
     }
   }
 
+  /**
+   * Releases a single resource based on containerId.
+   * @param containerId Yarn container ID
+   */
+  public void releaseResource(String containerId) {
+    if (StringUtils.isEmpty(containerId)) {
+      log.warn("ContainerId is not specified");
```
Comment (sborya) [inline samza-core/src/main/java/org/apache/samza/clustermanager/ResourceRequestState.java]: please add "ContainerId is not specified" 'in ClusterManager release resource'.
Without the context the message may be to vague.

**Pull request merged**